cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java

```diff
@@ -528,6 +542,86 @@ public void run() {
+      if (!Collection.class.isAssignableFrom(getShutdownHooksInOrder.getReturnType())) {
+        LOG.warn("Unsupported method {}. Spark shutdown hooks cleanup skipped.", getShutdownHooksInOrder);
+        return;
+      }
+      getShutdownHooksInOrder.setAccessible(true);
+
+      // Filter out hooks that are defined in the same SparkRunnerClassLoader as this SparkProgramRunner class
+      // This is for the case when there are concurrent Spark job running in the same VM
+      List<Runnable> hooks = ImmutableList.copyOf(
+        Iterables.filter(
+          Iterables.filter((Collection<?>) getShutdownHooksInOrder.invoke(manager), Runnable.class),
+          new Predicate<Runnable>() {
+            @Override
+            public boolean apply(Runnable runnable) {
+              return runnable.getClass().getClassLoader() == SparkRuntimeService.this.getClass().getClassLoader();
+            }
+          }
+        )
+      );
+
+      for (Runnable hook : hooks) {
+        LOG.debug("Running Spark shutdown hook {}", hook);
+        hook.run();
+        manager.removeShutdownHook(hook);
+      }
+
+    } catch (Exception e) {
+      LOG.warn("Failed to cleanup Spark shutdown hooks.", e);
+    }
+  }
+
+  /**
+   * Clear the constructor cache in the BeanIntrospector to avoid leaking ClassLoader.
+   */
+  private void invalidateBeanIntrospectorCache() {
+    try {
+      // Get the class through reflection since some Spark version doesn't depend on the fasterxml scala module.
+      // This is to avoid class not found issue
+      Class<?> cls = Class.forName("com.fasterxml.jackson.module.scala.introspect.BeanIntrospector$");
+      Field field = Fields.findField(cls, "ctorParamNamesCache");
+
+      // See if it is a LoadingCache. Need to check with class name since in distributed mode assembly jar, the
+      // guava classes are shaded and renamed.
+      switch (field.getType().getName()) {
+        // Get the cache field and invalid it.
+        // The BeanIntrospector is a scala object and scala generates a static MODULE$ field to the singleton object
+        case "com.google.common.cache.LoadingCache":
+        case "org.spark-project.guava.cache.LoadingCache":
+          field.setAccessible(true);
+          Object cache = field.get(Fields.findField(cls, "MODULE$").get(null));
+          // Try to call the invalidateAll method of the cache
+          Method invalidateAll = cache.getClass().getMethod("invalidateAll");
+          invalidateAll.setAccessible(true);
+          invalidateAll.invoke(cache);
+          LOG.debug("BeanIntrospector.ctorParamNamesCache is being invalidated.");
+        break;
+
+        default:
+          // Unexpected, maybe due to version change in the BeanIntrospector, hence log a WARN.
+          LOG.warn("BeanIntrospector.ctorParamNamesCache is not a LoadingCache, may lead to memory leak in SDK." +
+                     "Field type is {}", field.getType());
+      }
+    } catch (ClassNotFoundException e) {
```

Comment:
There is a comment in the next line. I updated the comment a bit to make it clearer.